include/reduce_helper.h

```diff
@@ -99,21 +112,38 @@ namespace quda
 #endif
     }
 
-    template <typename host_t, typename device_t = host_t> void complete(host_t *result, const qudaStream_t stream = 0)
+    template <typename host_t, typename device_t = host_t>
+    void complete(host_t *result, const qudaStream_t stream = 0, bool reset = false)
     {
+      if (launch_error == QUDA_ERROR) return; // kernel launch failed so return
+      if (launch_error == QUDA_ERROR_UNINITIALIZED) errorQuda("No reduction kernel appears to have been launched");
 #ifdef HETEROGENEOUS_ATOMIC
+      if (consumed) errorQuda("Cannot call complete more than once for each construction");
+
       for (int i = 0; i < n_reduce * n_item; i++) {
         result_h[i].wait(init_value<system_atomic_t>(), cuda::std::memory_order_relaxed);
       }
 #else
       auto event = reducer::get_event();
       qudaEventRecord(event, stream);
-      while (cudaSuccess != qudaEventQuery(event)) {}
+      while (!qudaEventQuery(event)) {}
 #endif
       // copy back result element by element and convert if necessary to host reduce type
       // unit size here may differ from system_atomic_t size, e.g., if doing double-double
       const int n_element = n_reduce * sizeof(T) / sizeof(device_t);
       for (int i = 0; i < n_element; i++) result[i] = reinterpret_cast<device_t *>(result_h)[i];
```

Comment:
I might be asking for too much here, but is it possible to have some kind of bound check here so we don't get untraceable setfault here? For example, having the interface as
```
void complete(std::vector<host_t> &result, const qudaStream_t stream = 0, bool reset = false)
```

Reply:
Done with https://github.com/lattice/quda/pull/1046/commits/e78a953461c1c38d76e116a3f9e8f3deffcb3e09